pdr_backend/data_eng/gql_data_factory.py

```diff
@@ -0,0 +1,267 @@
+       - "datetime" values ares python datetime.datetime, UTC
+    """
+
+    def __init__(self, pp: DataPP, ss: DataSS, web3: Web3PP):
+        self.pp = pp
+        self.ss = ss
+        self.web3 = web3
+
+        # TO DO: Solve duplicates from subgraph.
+        # Method 1: Cull anything returned outside st_ut, fin_ut
+        self.debug_duplicate = False
+
+        # TO DO: This code has DRY problems. Reduce.
+        # get network
+        if "main" in self.web3.network:
+            network = "mainnet"
+        elif "test" in self.web3.network:
+            network = "testnet"
+        else:
+            raise ValueError(self.web3.network)
+
+        # filter by feed contract address
+        contract_list = get_all_contract_ids_by_owner(
+            owner_address=self.web3.owner_addrs,
+            network=network,
+        )
+        contract_list = [f.lower() for f in contract_list]
+
+        # TO-DO: Roll into yaml config
+        self.record_config = {
+            "pdr_predictions": {
+                "fetch_fn": get_predictoor_predictions_df,
+                "schema": predictions_schema,
+                "config": {
+                    "contract_list": contract_list,
+                },
+            },
+        }
+
+    def get_gql_dfs(self) -> Dict[str, pl.DataFrame]:
+        """
+        @description
+          Get historical dataframes across many feeds and timeframes.
+
+        @return
+          predictions_df -- *polars* Dataframe. See class docstring
+        """
+        print("Get predictions data across many feeds and timeframes.")
+
+        # Ss_timestamp is calculated dynamically if ss.fin_timestr = "now".
+        # But, we don't want fin_timestamp changing as we gather data here.
+        # To solve, for a given call to this method, we make a constant fin_ut
+        fin_ut = self.ss.fin_timestamp
+
+        print(f"  Data start: {pretty_timestr(self.ss.st_timestamp)}")
+        print(f"  Data fin: {pretty_timestr(fin_ut)}")
+
+        self._update(fin_ut)
+        gql_dfs = self._load_parquet(fin_ut)
+
+        print("Get historical data across many subgraphs. Done.")
+
+        # postconditions
+        assert len(gql_dfs.values()) > 0
+        for df in gql_dfs.values():
+            assert isinstance(df, pl.DataFrame)
+
+        return gql_dfs
+
+    def _update(self, fin_ut: int):
+        """
+        @description
+            Iterate across all gql queries and update their parquet files:
+            - Predictoors
+            - Slots
+            - Claims
+
+            Improve this by:
+            1. Break out raw data from any transformed/cleaned data
+            2. Integrate other queries and summaries
+            3. Integrate config/pp if needed
+        @arguments
+            fin_ut -- a timestamp, in ms, in UTC
+        """
+
+        for k, record in self.record_config.items():
+            filename = self._parquet_filename(k)
```

Comment:
[data_factory common functions]
Hi, i disagree here.

every datafactory should have 3+ common fns, especially those that surround parquet
- save_parquet
- load_parquet
- parquet_filename

Although the class/object shouldn't be called ParquetDataFactory, the DataFactories should use these commons fns such that we can leverage common logic, and assert that every data_factory implements the same pattern

![image](https://github.com/oceanprotocol/pdr-backend/assets/69865342/7ba2cc5d-8b1f-44f0-a9fc-3b29ff664081)

Reply:
If it's  "gql data factory" that is "saving parquet", no reader will know what that data is. Because both "gql" and "parquet" are broad.

Please find a way to be more specific.

If you do this for "gql" then it might be enough.

Reply:
**[DataFactory Pattern]**
I believe we addressed this feedback inside of our standup.
- These common functions will be available inside all data_factories.
- The goal is to have gql_data_factory that yields multiple tables, in parquet format.
- The first table is called `pdr_predictions.parquet`

Next steps:
- Everything described here, will be further advanced inside #446 
- Updates to ohlcv_data_factory to rename functions will be made against yaml-cli2.

Resolving this conversation.